In [ ]:
!pip install pyarrow

In [ ]:
import pyarrow as pa

# Load the Arrow dataset file
arrow_file_path = "/content/drive/MyDrive/RARE25/Dataset/train/data-00000-of-00003.arrow"
with open(arrow_file_path, 'rb') as f:
    reader = pa.ipc.open_file(f)
    table = reader.read_all()

# Convert to Pandas DataFrame for easier manipulation
import pandas as pd
df = table.to_pandas()

# Check the first few rows of the DataFrame
print(df.head())


ArrowInvalid: Not an Arrow file

In [ ]:
with open(arrow_file_path, 'rb') as f:
    print(f.read(100))  # Print the first 100 bytes to inspect the file header

b'\xff\xff\xff\xff\xb0\x01\x00\x00\x10\x00\x00\x00\x00\x00\n\x00\x0e\x00\x06\x00\x05\x00\x08\x00\n\x00\x00\x00\x00\x01\x04\x00\x10\x00\x00\x00\x00\x00\n\x00\x0c\x00\x00\x00\x04\x00\x08\x00\n\x00\x00\x00\xac\x00\x00\x00\x04\x00\x00\x00\x01\x00\x00\x00\x0c\x00\x00\x00\x08\x00\x0c\x00\x04\x00\x08\x00\x08\x00\x00\x00\x80\x00\x00\x00\x04\x00\x00\x00q\x00\x00\x00{"info":'


In [ ]:
import pyarrow.parquet as pq

try:
    table = pq.read_table(arrow_file_path)
    print(table)
except Exception as e:
    print(f"Error reading as Parquet: {e}")


Error reading as Parquet: Error creating dataset. Could not read schema from '/content/drive/MyDrive/RARE25/Dataset/train/data-00000-of-00003.arrow'. Is this a 'parquet' file?: Could not open Parquet input source '/content/drive/MyDrive/RARE25/Dataset/train/data-00000-of-00003.arrow': Parquet magic bytes not found in footer. Either the file is corrupted or this is not a parquet file.


In [ ]:
# Read the first 1000 bytes to look for any patterns or recognizable headers
with open(arrow_file_path, 'rb') as f:
    header = f.read(1000)
    print(header)

b'\xff\xff\xff\xff\xb0\x01\x00\x00\x10\x00\x00\x00\x00\x00\n\x00\x0e\x00\x06\x00\x05\x00\x08\x00\n\x00\x00\x00\x00\x01\x04\x00\x10\x00\x00\x00\x00\x00\n\x00\x0c\x00\x00\x00\x04\x00\x08\x00\n\x00\x00\x00\xac\x00\x00\x00\x04\x00\x00\x00\x01\x00\x00\x00\x0c\x00\x00\x00\x08\x00\x0c\x00\x04\x00\x08\x00\x08\x00\x00\x00\x80\x00\x00\x00\x04\x00\x00\x00q\x00\x00\x00{"info": {"features": {"image": {"_type": "Image"}, "label": {"names": ["ndbe", "neo"], "_type": "ClassLabel"}}}}\x00\x00\x00\x0b\x00\x00\x00huggingface\x00\x02\x00\x00\x00@\x00\x00\x00\x04\x00\x00\x00p\xff\xff\xff\x00\x00\x01\x02\x10\x00\x00\x00 \x00\x00\x00\x04\x00\x00\x00\x00\x00\x00\x00\x05\x00\x00\x00label\x00\x00\x00\x08\x00\x0c\x00\x08\x00\x07\x00\x08\x00\x00\x00\x00\x00\x00\x01@\x00\x00\x00\xa8\xff\xff\xff\x00\x00\x01\r\x18\x00\x00\x00 \x00\x00\x00\x04\x00\x00\x00\x02\x00\x00\x00P\x00\x00\x00\x14\x00\x00\x00\x05\x00\x00\x00image\x00\x00\x00\xa0\xff\xff\xff\xd8\xff\xff\xff\x00\x00\x01\x05\x10\x00\x00\x00\x18\x00\x00\x00\x04\x0

In [ ]:
!pip install datasets

In [ ]:
from datasets import load_from_disk

try:
    # Assuming the .arrow file is part of a dataset structure that load_from_disk can recognize
    # We might need to point to the directory containing the file, not the file itself
    # Let's try the directory first
    dataset_dir = "/content/drive/MyDrive/RARE25/Dataset/train"
    dataset = load_from_disk(dataset_dir)
    print(dataset)
    # Convert to pandas for easier manipulation if needed
    df = dataset.to_pandas()
    print(df.head())

except Exception as e:
    print(f"Error loading dataset with datasets library: {e}")

Dataset({
    features: ['image', 'label'],
    num_rows: 3095
})
                                               image  label
0  {'bytes': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHD...      0
1  {'bytes': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHD...      0
2  {'bytes': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHD...      0
3  {'bytes': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHD...      0
4  {'bytes': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHD...      0


# Task
Extract images from the dataset at "/content/drive/MyDrive/RARE25/Dataset/train/data-00000-of-00003.arrow", save them to a separate folder, and create a CSV file containing the image file paths and corresponding labels.

## Create a directory for images

### Subtask:
Create a new directory to store the extracted images.


**Reasoning**:
Create a directory to store the extracted images using the `os` module.



In [ ]:
import os

image_dir = 'extracted_images'
os.makedirs(image_dir, exist_ok=True)

## Extract and save images

### Subtask:
Iterate through the DataFrame, extract the image bytes from each row, and save each image as a separate file in the created directory. Generate a filename for each image, perhaps based on its index or a unique identifier.


**Reasoning**:
Iterate through the DataFrame, extract the image bytes from each row, save each image as a separate file in the created directory, and generate a filename for each image.



In [ ]:
import io
from PIL import Image

for index, row in df.iterrows():
    image_bytes = row['image']['bytes']
    image = Image.open(io.BytesIO(image_bytes))
    filename = f'image_{index}.png'
    image_path = os.path.join(image_dir, filename)
    image.save(image_path)

print(f"Extracted and saved {len(df)} images to '{image_dir}'")

Extracted and saved 3095 images to 'extracted_images'


## Update dataframe with image paths

### Subtask:
Add a new column to the DataFrame containing the file paths where the images were saved.


**Reasoning**:
Create a list of image paths, add a new column 'image_path' to the dataframe with the image paths, and display the head of the updated dataframe to verify the changes.



In [ ]:
import os

image_dir = 'extracted_images'
image_paths = [os.path.join(image_dir, f'image_{index}.png') for index in df.index]
df['image_path'] = image_paths
display(df.head())

,image,label,image_path
0,{'bytes': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHD...,0,extracted_images/image_0.png
1,{'bytes': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHD...,0,extracted_images/image_1.png
2,{'bytes': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHD...,0,extracted_images/image_2.png
3,{'bytes': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHD...,0,extracted_images/image_3.png
4,{'bytes': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHD...,0,extracted_images/image_4.png


## Save dataframe to csv

### Subtask:
Save the updated DataFrame (with image paths and labels) to a CSV file.


**Reasoning**:
Save the updated DataFrame to a CSV file with the specified filename and without the index.



In [ ]:
output_csv_path = 'image_labels.csv'
df.to_csv(output_csv_path, index=False)
print(f"DataFrame saved to {output_csv_path}")

DataFrame saved to image_labels.csv


## Summary:

### Data Analysis Key Findings

*   A total of 3095 images were extracted from the dataset and saved as individual `.png` files in the `extracted_images` directory.
*   The filenames for the saved images were generated using the format `image_{index}.png`, where `{index}` corresponds to the row index in the original DataFrame.
*   A new column named `image_path` was successfully added to the DataFrame, containing the file path for each saved image.
*   The updated DataFrame, including the `image_path` and label columns, was saved to a CSV file named `image_labels.csv`.

### Insights or Next Steps

*   The created `image_labels.csv` file and the `extracted_images` directory can now be used for training a machine learning model or for further analysis.
*   Consider adding error handling for potential issues during image extraction or saving (e.g., corrupted image data).


In [ ]:
import os
import shutil

destination_folder = "/content/drive/MyDrive/RARE25/Dataset/extracted"

# Create the destination folder if it doesn't exist
os.makedirs(destination_folder, exist_ok=True)

# Define the source paths
image_dir = 'extracted_images'
output_csv_path = 'image_labels.csv'

# Define the destination paths
image_dir_dest = os.path.join(destination_folder, image_dir)
csv_dest_path = os.path.join(destination_folder, output_csv_path)

# Move the image directory
try:
    shutil.move(image_dir, image_dir_dest)
    print(f"Moved '{image_dir}' to '{image_dir_dest}'")
except FileNotFoundError:
    print(f"Error: Source directory '{image_dir}' not found.")
except Exception as e:
    print(f"Error moving directory: {e}")


# Move the CSV file
try:
    shutil.move(output_csv_path, csv_dest_path)
    print(f"Moved '{output_csv_path}' to '{csv_dest_path}'")
except FileNotFoundError:
    print(f"Error: Source file '{output_csv_path}' not found.")
except Exception as e:
    print(f"Error moving file: {e}")

Moved 'extracted_images' to '/content/drive/MyDrive/RARE25/Dataset/extracted/extracted_images'
Moved 'image_labels.csv' to '/content/drive/MyDrive/RARE25/Dataset/extracted/image_labels.csv'


In [ ]:
import pandas as pd
import os

# Define the path to the existing CSV file
destination_folder = "/content/drive/MyDrive/RARE25/Dataset/extracted"
output_csv_path = os.path.join(destination_folder, 'image_labels.csv')

# Read the existing CSV file into a DataFrame
try:
    df_existing = pd.read_csv(output_csv_path)
    print("Existing CSV file read successfully.")
    display(df_existing.head())
except FileNotFoundError:
    print(f"Error: The file '{output_csv_path}' was not found.")
    df_existing = None
except Exception as e:
    print(f"Error reading CSV file: {e}")
    df_existing = None

# If the DataFrame was loaded successfully, select the desired columns and save to a new CSV
if df_existing is not None:
    # Select only the 'image_path' and 'label' columns
    df_new_csv = df_existing[['image_path', 'label']].copy()

    # Define the path for the new CSV file
    new_output_csv_path = os.path.join(destination_folder, 'image_labels_simplified.csv')

    # Save the new DataFrame to a CSV file
    try:
        df_new_csv.to_csv(new_output_csv_path, index=False)
        print(f"New simplified CSV file saved successfully to '{new_output_csv_path}'")
        display(df_new_csv.head())
    except Exception as e:
        print(f"Error saving new CSV file: {e}")

Existing CSV file read successfully.


,image,label,image_path
0,{'bytes': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHD...,0,extracted_images/image_0.png
1,{'bytes': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHD...,0,extracted_images/image_1.png
2,{'bytes': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHD...,0,extracted_images/image_2.png
3,{'bytes': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHD...,0,extracted_images/image_3.png
4,{'bytes': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHD...,0,extracted_images/image_4.png


New simplified CSV file saved successfully to '/content/drive/MyDrive/RARE25/Dataset/extracted/image_labels_simplified.csv'


,image_path,label
0,extracted_images/image_0.png,0
1,extracted_images/image_1.png,0
2,extracted_images/image_2.png,0
3,extracted_images/image_3.png,0
4,extracted_images/image_4.png,0


In [ ]:
nic

# Task
Extract images and a CSV file containing image names and labels from the arrow file "/content/drive/MyDrive/RARE25/Dataset/train/data-00000-of-00003.arrow", save the extracted images to a folder named "extracted_images" and the CSV file to the specified Google Drive path "/content/drive/MyDrive/RARE25/Dataset/extracted", and then suggest a classification method to achieve higher PPV@90RECALL based on this data.

## Data preparation

### Subtask:
Load the simplified CSV with image paths and labels, and prepare the data for a machine learning model. This might involve splitting the data into training, validation, and testing sets.


**Reasoning**:
Load the simplified CSV into a DataFrame and then split it into training, validation, and testing sets with stratification.



In [ ]:
from sklearn.model_selection import train_test_split

# Load the simplified CSV file
destination_folder = "/content/drive/MyDrive/RARE25/Dataset/extracted"
new_output_csv_path = os.path.join(destination_folder, 'image_labels_simplified.csv')
df_simplified = pd.read_csv(new_output_csv_path)

# Split the data into training and testing sets (80% train, 20% test)
df_train, df_test = train_test_split(df_simplified, test_size=0.2, random_state=42, stratify=df_simplified['label'])

# Further split the training set into training and validation sets (80% train, 20% validation of the original train set)
df_train, df_val = train_test_split(df_train, test_size=0.25, random_state=42, stratify=df_train['label']) # 0.25 * 0.8 = 0.2 of original

# Print the shapes of the resulting DataFrames
print("Shape of training set:", df_train.shape)
print("Shape of validation set:", df_val.shape)
print("Shape of testing set:", df_test.shape)

Shape of training set: (1857, 2)
Shape of validation set: (619, 2)
Shape of testing set: (619, 2)


## Model selection

### Subtask:
Choose a suitable classification model. For image data, a Convolutional Neural Network (CNN) is often a good starting point. Consider transfer learning with a pre-trained model.


## Model selection

### Subtask:
Choose a suitable classification model. For image data, a Convolutional Neural Network (CNN) is often a good starting point. Consider transfer learning with a pre-trained model.


## Model training

### Subtask:
Train a suitable classification model (e.g., a CNN with transfer learning) on the training data (`df_train`).


**Reasoning**:
Train a CNN model using transfer learning.



In [ ]:
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.applications import VGG16
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.models import Model
import numpy as np
from PIL import Image

# Define image dimensions and number of classes
img_height, img_width = 128, 128  # Adjust based on typical image size or desired input size
num_classes = 2  # Assuming binary classification based on the label column

# Create data generators
train_datagen = ImageDataGenerator(
    rescale=1./255,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True
)

val_datagen = ImageDataGenerator(rescale=1./255)

# Create generators from DataFrame
train_generator = train_datagen.flow_from_dataframe(
    dataframe=df_train,
    directory="/content/drive/MyDrive/RARE25/Dataset/extracted", # Directory containing the images
    x_col="image_path",
    y_col="label",
    target_size=(img_height, img_width),
    batch_size=32,
    class_mode="raw" # Use "raw" for numerical labels
)

validation_generator = val_datagen.flow_from_dataframe(
    dataframe=df_val,
    directory="/content/drive/MyDrive/RARE25/Dataset/extracted", # Directory containing the images
    x_col="image_path",
    y_col="label",
    target_size=(img_height, img_width),
    batch_size=32,
    class_mode="raw" # Use "raw" for numerical labels
)

# Load pre-trained VGG16 model
base_model = VGG16(weights='imagenet', include_top=False, input_shape=(img_height, img_width, 3))

# Add new layers for classification
x = base_model.output
x = Flatten()(x)
x = Dense(256, activation='relu')(x)
predictions = Dense(num_classes, activation='softmax')(x) # Use softmax for multi-class, sigmoid for binary

# Create the final model
model = Model(inputs=base_model.input, outputs=predictions)

# Freeze the layers of the pre-trained model
for layer in base_model.layers:
    layer.trainable = False

# Compile the model
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy']) # Use sparse_categorical_crossentropy for integer labels

# Train the model
history = model.fit(
    train_generator,
    steps_per_epoch=train_generator.samples // train_generator.batch_size,
    epochs=10, # Adjust number of epochs
    validation_data=validation_generator,
    validation_steps=validation_generator.samples // validation_generator.batch_size
)

Found 1857 validated image filenames.
Found 619 validated image filenames.
58889256/58889256 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


/usr/local/lib/python3.11/dist-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


Epoch 1/10
58/58 ━━━━━━━━━━━━━━━━━━━━ 552s 10s/step - accuracy: 0.9472 - loss: 0.6100 - val_accuracy: 0.9507 - val_loss: 0.2075
Epoch 2/10
 1/58 ━━━━━━━━━━━━━━━━━━━━ 7:21 8s/step - accuracy: 1.0000 - loss: 0.0122

/usr/local/lib/python3.11/dist-packages/keras/src/trainers/epoch_iterator.py:107: UserWarning: Your input ran out of data; interrupting training. Make sure that your dataset or generator can generate at least `steps_per_epoch * epochs` batches. You may need to use the `.repeat()` function when building your dataset.
  self._interrupted_warning()


58/58 ━━━━━━━━━━━━━━━━━━━━ 143s 2s/step - accuracy: 1.0000 - loss: 0.0122 - val_accuracy: 0.9490 - val_loss: 0.2349
Epoch 3/10
58/58 ━━━━━━━━━━━━━━━━━━━━ 598s 10s/step - accuracy: 0.9503 - loss: 0.1908 - val_accuracy: 0.9507 - val_loss: 0.2962
Epoch 4/10
58/58 ━━━━━━━━━━━━━━━━━━━━ 144s 2s/step - accuracy: 0.9375 - loss: 0.2711 - val_accuracy: 0.9490 - val_loss: 0.2933
Epoch 5/10
58/58 ━━━━━━━━━━━━━━━━━━━━ 628s 10s/step - accuracy: 0.9505 - loss: 0.1545 - val_accuracy: 0.9490 - val_loss: 0.1838
Epoch 6/10
58/58 ━━━━━━━━━━━━━━━━━━━━ 141s 2s/step - accuracy: 0.9688 - loss: 0.1100 - val_accuracy: 0.9490 - val_loss: 0.1818
Epoch 7/10
58/58 ━━━━━━━━━━━━━━━━━━━━ 559s 10s/step - accuracy: 0.9506 - loss: 0.1632 - val_accuracy: 0.9507 - val_loss: 0.2654
Epoch 8/10
58/58 ━━━━━━━━━━━━━━━━━━━━ 142s 2s/step - accuracy: 0.9375 - loss: 0.1506 - val_accuracy: 0.9490 - val_loss: 0.2770
Epoch 9/10
58/58 ━━━━━━━━━━━━━━━━━━━━ 624s 10s/step - accuracy: 0.9567 - loss: 0.1526 - val_accuracy: 0.9523 - val_loss

## Model evaluation

### Subtask:
Evaluate the trained model on the validation set, focusing on metrics like precision, recall, and F1-score. Specifically, analyze the precision at the desired recall level (90%).


**Reasoning**:
Evaluate the trained model on the validation set and calculate precision at 90% recall.



In [ ]:
import numpy as np
from sklearn.metrics import classification_report, precision_recall_curve

# Get true labels and predicted probabilities for the validation set
# We need to reset the generator to ensure we get predictions for the entire validation set
validation_generator.reset()
y_true = validation_generator.classes
y_pred_probs = model.predict(validation_generator)

# Calculate predicted class labels
y_pred = np.argmax(y_pred_probs, axis=1)

# Calculate and print the classification report
print("Classification Report on Validation Set:")
print(classification_report(y_true, y_pred))

# Calculate precision and recall for various thresholds
precision, recall, thresholds = precision_recall_curve(y_true, y_pred_probs[:, 1]) # Assuming class 1 is the positive class

# Find the precision at approximately 90% recall
# Iterate through the recall values and find the precision where recall is close to 0.9
precision_at_90_recall = None
for i in range(len(recall)):
    if recall[i] >= 0.9:
        precision_at_90_recall = precision[i]
        break

# Print the precision at 90% recall
print(f"Precision at approximately 90% Recall: {precision_at_90_recall}")

**Reasoning**:
The previous code failed because `DataFrameIterator` does not have a `classes` attribute. We need to get the true labels from the original validation DataFrame instead of the generator. We also need to get all predictions by iterating through the generator, not just calling `model.predict` once.



In [ ]:
import numpy as np
from sklearn.metrics import classification_report, precision_recall_curve

# Get true labels from the original validation DataFrame
y_true = df_val['label'].values

# Get predicted probabilities for the validation set by iterating through the generator
validation_generator.reset()
y_pred_probs_list = []
for i in range(validation_generator.samples // validation_generator.batch_size + (validation_generator.samples % validation_generator.batch_size > 0)):
    images, labels = next(validation_generator)
    preds = model.predict(images)
    y_pred_probs_list.append(preds)

y_pred_probs = np.concatenate(y_pred_probs_list, axis=0)

# Calculate predicted class labels
y_pred = np.argmax(y_pred_probs, axis=1)

# Calculate and print the classification report
print("Classification Report on Validation Set:")
print(classification_report(y_true, y_pred))

# Calculate precision and recall for various thresholds
precision, recall, thresholds = precision_recall_curve(y_true, y_pred_probs[:, 1]) # Assuming class 1 is the positive class

# Find the precision at approximately 90% recall
# Iterate through the recall values and find the precision where recall is close to 0.9
precision_at_90_recall = None
for i in range len(recall):
    if recall[i] >= 0.9:
        precision_at_90_recall = precision[i]
        break

# Print the precision at 90% recall
print(f"Precision at approximately 90% Recall: {precision_at_90_recall}")

SyntaxError: invalid syntax (ipython-input-2899220517.py, line 30)

**Reasoning**:
The previous code had a syntax error in the `for` loop for finding the precision at 90% recall. The `range` function needs parentheses around its argument. I will fix the syntax error and re-run the code to evaluate the model and calculate the precision at 90% recall.



In [ ]:
import numpy as np
from sklearn.metrics import classification_report, precision_recall_curve

# Get true labels from the original validation DataFrame
y_true = df_val['label'].values

# Get predicted probabilities for the validation set by iterating through the generator
validation_generator.reset()
y_pred_probs_list = []
for i in range(validation_generator.samples // validation_generator.batch_size + (validation_generator.samples % validation_generator.batch_size > 0)):
    images, labels = next(validation_generator)
    preds = model.predict(images)
    y_pred_probs_list.append(preds)

y_pred_probs = np.concatenate(y_pred_probs_list, axis=0)

# Calculate predicted class labels
y_pred = np.argmax(y_pred_probs, axis=1)

# Calculate and print the classification report
print("Classification Report on Validation Set:")
print(classification_report(y_true, y_pred))

# Calculate precision and recall for various thresholds
precision, recall, thresholds = precision_recall_curve(y_true, y_pred_probs[:, 1]) # Assuming class 1 is the positive class

# Find the precision at approximately 90% recall
# Iterate through the recall values and find the precision where recall is close to 0.9
precision_at_90_recall = None
for i in range(len(recall)):
    if recall[i] >= 0.9:
        precision_at_90_recall = precision[i]
        break

# Print the precision at 90% recall
print(f"Precision at approximately 90% Recall: {precision_at_90_recall}")

1/1 ━━━━━━━━━━━━━━━━━━━━ 6s 6s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 8s 8s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 6s 6s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 6s 6s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 6s 6s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 7s 7s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 6s 6s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 6s 6s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 6s 6s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 6s 6s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 7s 7s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 6s 6s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 8s 8s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 6s 6s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 6s 6s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 8s 8s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 6s 6s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 7s 7s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 6s 6s/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 3s 3s/step
Classification Report on Validation Set:
              precision    recall  f1-score   support

           0       0.95      1.00      0.97       588
           1       0.00      0.00      0.00        31

    accuracy                           0.95       619
   macro avg       0.

## Hyperparameter tuning and optimization

### Subtask:
Fine-tune the model's hyperparameters or explore different model architectures to improve performance, with a focus on increasing PPV at 90% recall. This might involve adjusting the classification threshold.


**Reasoning**:
Analyze the previous results and determine strategies to improve PPV at 90% recall, specifically for the minority class. The classification report showed very low performance for class 1. To address this, I will try adjusting the classification threshold. I will also implement class weighting in the model training to give more importance to the minority class during training. I will then retrain the model with these adjustments.



In [ ]:
import pandas as pd
import os

# Define the path to the simplified CSV file
destination_folder = "/content/drive/MyDrive/RARE25/Dataset/extracted"
simplified_csv_path = os.path.join(destination_folder, 'image_labels_simplified.csv')

# Read the simplified CSV file into a DataFrame
try:
    df_analysis = pd.read_csv(simplified_csv_path)
    print("Simplified CSV file loaded successfully for analysis.")
    display(df_analysis.head())

    # Display basic statistics
    print("\nDataset Information:")
    df_analysis.info()

    print("\nValue Counts for Labels:")
    display(df_analysis['label'].value_counts())

except FileNotFoundError:
    print(f"Error: The file '{simplified_csv_path}' was not found.")
except Exception as e:
    print(f"Error reading CSV file for analysis: {e}")

Simplified CSV file loaded successfully for analysis.


,image_path,label
0,extracted_images/image_0.png,0
1,extracted_images/image_1.png,0
2,extracted_images/image_2.png,0
3,extracted_images/image_3.png,0
4,extracted_images/image_4.png,0



Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   image_path  3095 non-null   object
 1   label       3095 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 48.5+ KB

Value Counts for Labels:


,count
label,
0,2937
1,158
